### PR TITLE
feat(components): add sass build pipeline

### LINKED ISSUE #4001: Add build pipeline for `carbon-components` Sass files for CSS Custom Property fallback
We're looking to add support for CSS Custom Property to our source files. This format will follow:

```diff
.#{$prefix}--component-name {
-  color: $token-01;
+  color: var(--token-01, $token-01);
}
```

However, we can't ship CSS Custom Properties as a direct replacement as this could break existing users during a minor bump. As a result, we need to introduce a build pipeline that will take in our source sass files and do the following transform:

```scss
// Input
.#{$prefix}--component-name {
  color: var(--token-01, $token-01);
}

// Output
.#{$prefix}--component-name {
  color: $token-01;
  color: var(--token-01, $token-01);
}
```

Thankfully, we can use a combination of `postcss`, `postcss-scss`, and `postcss-css-variables` to accomplish this for our output `scss` directory.

This will need to be introduced in the `sass:source` gulp task:

```js
gulp.task('sass:source', () =>
  gulp.src('./src/**/*.scss').pipe(gulp.dest('scss'))
);
```

Most likely with the following modification:

```js
const cssVariables = require('postcss-css-variables');
const scss = require('postcss-scss');

gulp.task('sass:source', () =>
  gulp
    .src('./src/**/*.scss')
    .pipe(postcss([cssVariables({ preserve: true })], { parsers: scss }))
    .pipe(gulp.dest('scss'))
);
```

_Note: we use `postcss-css-variables` instead of the alternative as it can handle `$token-01` as a value type_

### PR DESCRIPTION
**Pull request by joshblack**
Closes #4001

Adds in a build step for our `scss` folder where we transform custom property usage with an appropriate fallback.

#### Changelog

**New**

- Custom postcss plugin to transform CSS Custom Properties into Sass variable fallbacks

**Changed**

- `gulp sass:source` now uses postcss and a custom plugin to transform source files to include an appropriate fallback

**Removed**

#### Testing / Reviewing

- Pull down PR
- Run `yarn build` in `components`
- View `scss/components/_accordion.scss` and verify this output:

```scss
/*/ Accordion styles*/
/*/ @access private*/
/*/ @group accordion*/
@mixin accordion {
  .#{$prefix}--accordion {
    @include reset;

    list-style: none;
    width: 100%;
  }

  .#{$prefix}--accordion__item {
    transition: all $duration--fast-02 motion(standard, productive);
    transition: all var(--c-duration-fast-02, $duration--fast-02)
      motion(standard, productive);
    border-top: 1px solid $ui-03;
    overflow: visible;

    &:last-child {
      border-bottom: 1px solid $ui-03;
      border-bottom: 1px solid var(--c-ui-03, $ui-03);
    }
  }

```

### CODE REVIEW

**Comment by vpicone** [comment]
@joshblack I think adding a custom postcss plugin adds a lot of complexity that will be tough for others to grok. I was able to get the same out put using [post-css-variables](https://github.com/MadLittleMods/postcss-css-variables) with the `preserve` option turned on. They have a [playground](https://madlittlemods.github.io/postcss-css-variables/playground/) where you can test input.


I had to take out our motion function since this doesn't use the scss parser:
Screen Shot 2019-09-18 at 11 39 10 AM


There's a few [features](https://github.com/MadLittleMods/postcss-css-variables#pseudo-classes-and-elements) handled by the plugin I'm not sure we cover here: media queries/pseudo selectors redefining variables used by the parent, and nested rules.

**Comment by joshblack** [comment]
@vpicone responded on slack, to repeat it here I definitely went through the common plugins with no success 😞 unfortunately each one had some issue associated with it.

The one above you mentioned will give this error: 
image

**Comment by joshblack** [comment]
Note: if we get more people +1'ing #4030 then this will be unnecessary

File: packages/components/tools/postcss/transform-custom-properties-test.js [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @jest-environment node
+ */
+
+'use strict';
+
+const postcss = require('postcss');
+const scss = require('postcss-scss');
```
Comment (emyarod) [inline packages/components/tools/postcss/transform-custom-properties-test.js:13]: are we missing `postcss-scss` in the devDependencies?

**Review by emyarod — COMMENTED** [review]
still writing some test cases but here is one test I thought of that is currently failing (assuming I wrote it correctly), wanted to document this first before I finish my review

```js
{
    name: 'multiline properties',
    input: `
.#{$prefix}--component {
  background: linear-gradient(
    var(--val-01, $val-01),
    var(--val-02, $val-02),
    orange 30%,
    yellow 50%,
    yellow 70%,
    green 90%,
    green 100%
  );
  background: linear-gradient(var(--val-01, $val-01), var(--val-02, $val-02) 30%, var(--val-03, $val-03), green 90% 100%);
}
`,
    output: `
.#{$prefix}--component {
  background: linear-gradient(
    $val-01,
    $val-02,
    orange 30%,
    yellow 50%,
    yellow 70%,
    green 90%,
    green 100%
  );
  background: linear-gradient(
    var(--val-01, $val-01),
    var(--val-02, $val-02),
    orange 30%,
    yellow 50%,
    yellow 70%,
    green 90%,
    green 100%
  );
  background: linear-gradient($val-01, $val-02 30%, $val-03, green 90% 100%);
  background: linear-gradient(var(--val-01, $val-01), var(--val-02, $val-02) 30%, var(--val-03, $val-03), green 90% 100%);
}
`,
  }
```

**Comment by joshblack** [comment]
Closing based on slack discussion 👍

**Pull request closed without merging**